### PR TITLE
Keccak

### PR DESCRIPTION
**Pull request by chriseth**
Implements #82

### CODE REVIEW

File: tests/assert_pass/keccak256_generic.yul [hunk 1]
```diff
@@ -0,0 +1,9 @@
+{
+    mstore(0, calldataload(0))
+    let x := keccak256(0, calldataload(7))
+//    mstore(0, calldataload(9))
```
Comment (chriseth) [inline tests/assert_pass/keccak256_generic.yul:4]: Without this the test is failing, although the condition should actually ensure that the memory contents are the same, isn't it?

File: src/evm_context.rs [hunk 2]
```diff
@@ -38,6 +39,26 @@ pub fn init(ssa: &mut SSATracker) -> String {
...
+        output,
+        "(declare-fun {} ((_ BitVec 256) (_ BitVec 256)) (_ BitVec 256))",
+        keccak256_64()
+    );
+
+    let _ = writeln!(
+        output,
+        "(declare-fun {} ({} (_ BitVec 256) (_ BitVec 256)) (_ BitVec 256))",
+        keccak256_generic(),
+        memory_type(),
```
Comment (leonardoalt) [inline src/evm_context.rs:59]: that's not really correct, is it?

File: src/evm_context.rs [hunk 3]
```diff
@@ -168,6 +189,26 @@ pub fn mload(index: &String, ssa: &mut SSATracker) -> String {
     format!("(concat {})", arguments)
 }
 
+pub fn keccak256(offset: &String, length: &String, ssa: &mut SSATracker) -> String {
+    let scratch_0 = mload(&smt::SMTFormat::as_smt(&0), ssa);
+    let scratch_1 = mload(&smt::SMTFormat::as_smt(&0x20), ssa);
+    smt::ite(
+        &smt::and(&smt::eq(offset, &0), &smt::eq(length, &0x20)),
```
Comment (leonardoalt) [inline src/evm_context.rs:196]: Do we need to check whether offset is 0? Could also just `mload` from whatever offset is as long as the length is correct, and hope the SMT solver's constant propagation takes care of the 0 here

Comment (chriseth) [reply src/evm_context.rs:196]: Oh right, that is totally true!

File: tests/revert_unreachable/keccak256_32.yul [hunk 4]
```diff
@@ -0,0 +1,15 @@
+{
+    mstore(0, calldataload(0))
+    mstore(0x20, calldataload(1))
+    let x := keccak256(0, 32)
+    mstore(64, calldataload(99))
```
Comment (leonardoalt) [inline tests/revert_unreachable/keccak256_32.yul:5]: Nit, but maybe stick to hex in the memory addresses, easier to read

Comment (leonardoalt) [inline tests/revert_unreachable/keccak256_32.yul:5]:
```suggestion
    mstore(0x40, calldataload(99))
```

File: tests/revert_unreachable/keccak256_32.yul [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+    let y := keccak256(0, 32)
+    mstore(0, calldataload(100))
+    mstore(0x20, calldataload(200))
+    if and(
+        eq(calldataload(0), calldataload(100)),
+        eq(calldataload(1), calldataload(200))
+    ) {
+        if iszero(eq(x, y)) { revert(0, 0 )}
```
Comment (leonardoalt) [inline tests/revert_unreachable/keccak256_32.yul:13]: Nice!

File: src/evm_context.rs [hunk 6]
```diff
@@ -218,6 +259,18 @@ fn memory_type() -> &'static str {
     "(Array (_ BitVec 256) (_ BitVec 8))"
 }
 
+fn keccak256_32() -> String {
+    "_keccak256_32".to_string()
+}
+
+fn keccak256_64() -> String {
+    "_keccak256_32".to_string()
```
Comment (leonardoalt) [inline src/evm_context.rs:267]:
```suggestion
    "_keccak256_64".to_string()
```

Comment (leonardoalt) [reply src/evm_context.rs:267]: This is the only thing that needs real fixing.

File: src/evm_context.rs [hunk 7]
```diff
@@ -168,6 +189,26 @@ pub fn mload(index: &String, ssa: &mut SSATracker) -> String {
...
 
+pub fn keccak256(offset: &String, length: &String, ssa: &mut SSATracker) -> String {
+    let offset_0 = mload(offset, ssa);
+    let offset_32 = mload(&format!("(bvadd {} #x{:064X})", offset, 0x20), ssa);
+    smt::ite(
+        &smt::eq(length, &0x20),
+        &format!("({} {})", keccak256_32(), offset_0),
+        &smt::ite(
+            &smt::eq(length, &0x40),
+            &format!("({} {} {})", keccak256_32(), offset_0, offset_32),
```
Comment (chriseth) [inline src/evm_context.rs:200]:
```suggestion
            &format!("({} {} {})", keccak256_64(), offset_0, offset_32),
```
:)

Comment (chriseth) [reply src/evm_context.rs:200]: Why did this not result in an smt parser error?

Comment (leonardoalt) [reply src/evm_context.rs]: good q, can you check that the tests are covering this?

**Review by leonardoalt — APPROVED** [review]

**Pull request merged**